app/esper/rekall_vgrid.py

```diff
@@ -0,0 +1,354 @@
+
+VGrid displays a list of VBlocks. Each VBlock can have multiple tracks, and
+each track shows a timeline of intervals in that track, as well as drawing
+objects and showing metadata of the interval as configured.
+
+There are two main use cases of VGrid:
+    1. VideoVBlocks. This mode displays a VBlock for each video. Each track
+        in the VBlock is an IntervalSet to visualize in that video.
+    2. IntervalVBlocks. This mode displays a VBlock for each interval/clip.
+        Each track in the VBlock is an IntervalSet to visualize in that
+        interval/clip.
+
+The VBlocksBuilders in this module helps building the JSON structure for this
+list of VBlocks.
+
+Example of using VideoVBlock:
+
+    # face_collection is a DomainIntervalCollection of face intervals.
+
+    # We want to see a list of VBlocks where each is a video in the collection
+    # We want to see two tracks in each VBlock: one showing all faces in the
+    # collection with bounding boxes drawn and a flag metadata set; another
+    # showing all captions of the video.
+
+    json = VideoVBlocksBuilder()\\
+        .add_track(
+            VideoTrackBuilder('faces', face_collection)\\
+                .set_draw_type(DrawType_Bbox())\\
+                .add_metadata('flag', Metadata_Flag()))\\
+        .add_track(
+            CaptionTrack())\\
+        .build()
+
+Example of using IntervalBlock:
+
+    # conversations_with_ids is a DomainIntervalCollection of conversation
+    # sequences, where the payload on each conversation is an IntervalSet of
+    # face bounding boxes with identity as payload.
+
+    # We want to see a list of VBlocks where each is a conversation 
+    # sequence in the collection. We want to see two tracks in each VBlock:
+    # one showing the conversation sequence; another showing all the faces with
+    # the bounding boxes drawn and generic metadata showing the identities.
+
+    json = IntervalVBlocksBuilder()\\
+        .add_track(
+            IntervalTrackBuilder('conversation'))\\
+        .add_track(
+            IntervalTrackBuilder('faces', lambda i:i.payload)\\
+                    .set_draw_type(DrawType_Bbox())\\
+                    .add_metadata('identity', Metadata_Generic()))\\
+        .build(conversations_with_ids)
+
+Symbols exported in this module are the following.
+
+VBlocksBuilders:
+    VideoVBlocksBuilder: VideoVBlocks mode.
+    IntervalVBlocksBuilder: IntervalVBlocks mode.
+
+Tracks:
+    VideoTrackBuilder: Build custom tracks for VideoVBlocks mode.
+    IntervalTrackBuilder: Build custom tracks for IntervalVBlocks mode.
+    CaptionTrack: Add caption to the VBlocks. Works in either mode.
+
+DrawTypes:
+    DrawType_Bbox: Draw bounding boxes around the spatial dimension.
+    DrawType_Caption: Display text in the caption box.
+
+Metadatas:
+    Metadata_Flag: Plant a flag marking the temporal interval.
+    Metadata_Generic: Show a generic JSON text.
+"""
```

Comment:
Can you show an example that goes from database -> videovblock? I think it's important that the whole workflow should be small. Let's say we want to visualize faces + genders from FOX.

Reply:
I still don't have a good understanding of what `database` is, and how to use that for `Metadata_Categorical` and `Category`. In VGrid code I only see `database` used in video_id lookup which is not related to this module.

Reply:
Broadly the goal of `database` is to be "all rows from the SQL database containing metadata about objects referenced in the provided intervalsets." Right now that's just video metadata, but could include other things in the future.

I haven't fully decided on the right way to handle categories, that API isn't fully implemented. Going to figure that out this weekend so don't worry about it for now.